shared-module/_pixelbuf/PixelBuf.c

```diff
@@ -26,95 +26,256 @@
 
 
 #include "py/obj.h"
-#include "py/objarray.h"
+#include "py/objstr.h"
+#include "py/objtype.h"
 #include "py/runtime.h"
-#include "PixelBuf.h"
+#include "shared-bindings/_pixelbuf/PixelBuf.h"
 #include <string.h>
 
-void pixelbuf_set_pixel_int(uint8_t *buf, mp_int_t value, pixelbuf_byteorder_details_t *byteorder) {
-    buf[byteorder->byteorder.r] = value >> 16 & 0xff;
-    buf[byteorder->byteorder.g] = (value >> 8) & 0xff;
-    buf[byteorder->byteorder.b] = value & 0xff;
-    if (byteorder->bpp == 4 && byteorder->has_white && 
-            (buf[byteorder->byteorder.r] == buf[byteorder->byteorder.g] && 
-                buf[byteorder->byteorder.r] == buf[byteorder->byteorder.b])) {
-        buf[byteorder->byteorder.w] = buf[byteorder->byteorder.r];
-        buf[byteorder->byteorder.r] = buf[byteorder->byteorder.g] = buf[byteorder->byteorder.b] = 0;
-    }
-}
-
-void pixelbuf_set_pixel(uint8_t *buf, uint8_t *rawbuf, float brightness, mp_obj_t *item, pixelbuf_byteorder_details_t *byteorder, bool dotstar) {
-    if (MP_OBJ_IS_INT(item)) {
-        uint8_t *target = rawbuf ? rawbuf : buf;
-        pixelbuf_set_pixel_int(target, mp_obj_get_int_truncated(item), byteorder);
-        if (dotstar) {
-            buf[0] = DOTSTAR_LED_START_FULL_BRIGHT;
-            if (rawbuf) 
-                rawbuf[0] = DOTSTAR_LED_START_FULL_BRIGHT;
+// Helper to ensure we have the native super class instead of a subclass.
+static pixelbuf_pixelbuf_obj_t* native_pixelbuf(mp_obj_t pixelbuf_obj) {
+    mp_obj_t native_pixelbuf = mp_instance_cast_to_native_base(pixelbuf_obj, &pixelbuf_pixelbuf_type);
+    mp_obj_assert_native_inited(native_pixelbuf);
+    return MP_OBJ_TO_PTR(native_pixelbuf);
+}
+
+void common_hal__pixelbuf_pixelbuf_construct(pixelbuf_pixelbuf_obj_t *self, size_t n,
+        pixelbuf_byteorder_details_t* byteorder, mp_float_t brightness, bool auto_write,
+        uint8_t* header, size_t header_len, uint8_t* trailer, size_t trailer_len) {
+
+    self->pixel_count = n;
+    self->byteorder = *byteorder;  // Copied because we modify for dotstar
+    self->bytes_per_pixel = byteorder->is_dotstar ? 4 : byteorder->bpp;
+    self->auto_write = false;
+
+    size_t pixel_len = self->pixel_count * self->bytes_per_pixel;
+    self->transmit_buffer_obj = mp_obj_new_bytes_of_zeros(header_len + pixel_len + trailer_len);
+    mp_obj_str_t *o = MP_OBJ_TO_PTR(self->transmit_buffer_obj);
+
+    // Abuse the bytes object a bit by mutating it's data by dropping the const. If the user's
+    // Python code holds onto it, they'll find out that it changes. At least this way it isn't
+    // mutable by the code itself.
+    uint8_t* transmit_buffer = (uint8_t*) o->data;
+    memcpy(transmit_buffer, header, header_len);
+    memcpy(transmit_buffer + header_len + pixel_len, trailer, trailer_len);
```

Comment:
Correct me if I'm wrong, but doesn't this make the header and trailer fixed at PixelBuf construction?   While this is fine for Dotstar, what if some other protocol needs to manipulate the header or trailer?  (Not that this is necessarily a bad thing).

Reply:
It does fix it. We can deal with other protocols when we find one that needs it.